stdlib/ssl.pyi

```diff
@@ -316,8 +316,8 @@ class _ASN1Object(_ASN1ObjectBase):
     def fromname(cls, name: str) -> Self: ...
 
 class Purpose(_ASN1Object, enum.Enum):
-    SERVER_AUTH: _ASN1Object
-    CLIENT_AUTH: _ASN1Object
+    SERVER_AUTH = (129, "serverAuth", "TLS Web Server Authentication", "1.3.6.1.5.5.7.3.2")  # pyright: ignore[reportCallIssue]
+    CLIENT_AUTH = (130, "clientAuth", "TLS Web Client Authentication", "1.3.6.1.5.5.7.3.1")  # pyright: ignore[reportCallIssue]
```

Comment:
What was pyright complaining about here, out of curiosity?

Reply:
When an enum instance is constructed and a tuple value is provided (like in this example), the tuple is expanded and passed to the constructor as individual positional arguments. In this case, the `__new__` and `__init__` methods should accept four arguments. The `__new__` and `__init__` methods for the `_ASN1ObjectBase` (which are auto-generated by the NamedTuple mechanism) do accept four arguments. But the `_ASN1Object.__new__` method accepts only a single positional argument called "oid". Pyright is noting this problem. This looks like a bug in the type stub because this code will not run as it's currently defined in the stub. Try pasting the following code into a ".py" file and running it.

```python
from enum import Enum
from typing import NamedTuple, Self

class _ASN1ObjectBase(NamedTuple):
    nid: int
    shortname: str
    longname: str
    oid: str

class _ASN1Object(_ASN1ObjectBase):
    def __new__(cls, oid: str) -> Self: ...
    @classmethod
    def fromnid(cls, nid: int) -> Self: ...
    @classmethod
    def fromname(cls, name: str) -> Self: ...

class Purpose(_ASN1Object, Enum):
    SERVER_AUTH = (
        129,
        "serverAuth",
        "TLS Web Server Authentication",
        "1.3.6.1.5.5.7.3.2",
    )
    CLIENT_AUTH = (
        130,
        "clientAuth",
        "TLS Web Client Authentication",
        "1.3.6.1.5.5.7.3.1",
    )
```

Reply:
Interesting -- thanks for the explanation. I don't think there's a pre-existing bug in the stubs here -- the signatures look accurate to me. The runtime code is here, and is pretty readable: https://github.com/python/cpython/blob/c06be6bbb8d138dde50c0a07cbd64496bee537c5/Lib/ssl.py#L394-L419

Rather, the issue is that what the runtime is doing is pretty dynamic, and hard to represent statically in a a stub file. The runtime is assigning a single string value to each enum member in the class scope, but that string value is then passed to `_ASN1Object.__new__` as part of the creation of the enum class. `_ASN1Object.__new__` "expands" the single string value into a four-item tuple, which then becomes the value of the final enum member at runtime.